rfcs/0006-merge-operator.md

```diff
@@ -0,0 +1,144 @@
+# SlateDB Merge Operator
+
+Status: Draft
+
+Authors:
+
+* [David Moravek](https://github.com/dmvk)
+
+References:
+
+* https://github.com/facebook/rocksdb/wiki/Merge-Operator
+* https://github.com/facebook/rocksdb/wiki/Merge-Operator#associativity-vs-non-associativity
+* https://github.com/dgraph-io/badger/blob/v4.4.0/merge.go
+
+## Motivation
+
+To utilize SlateDB as a fully-featured state backend for a stream processor, it must support three key types of state:
+
+1. **Value**: Latest value or aggregating state.
+2. **List / Bag / Buffering**: State that stores multiple items for later processing.
+3. **Map**: Key-value state for fine-grained data.
+
+State-of-the-art implementations built on RocksDB often leverage the [Merge Operator](https://github.com/facebook/rocksdb/wiki/Merge-Operator) to efficiently manage types 1 and 2. A prime example is list state, where elements are appended continuously until triggered by an event like a window firing. While this method may not always be the absolute optimal solution, it serves as a solid foundation for future iterations and enhancements.
+
+## Goals
+The primary goal is to introduce a new user-facing record type (**Merge Operand**) and a **Merge Operator** interface to SlateDB, allowing applications to bypass the traditional read/modify/update cycle in performance-critical situations where computation can be expressed using an [associative](https://github.com/facebook/rocksdb/wiki/Merge-Operator#associativity-vs-non-associativity) operator.
+
+**Supporting non-associative operators is not a goal of this RFC**. The assumption is that associative operators can cover the majority of use cases while keeping the user-facing API simple and significantly reducing the scope. [Badger](https://github.com/dgraph-io/badger/blob/v4.4.0/merge.go) has adopted the same limitation.
+
+## Public API
```

Comment:
This is compatible with this RFC, so we can (and probably should) leave it as a follow up, but wanted to drop this note anyway. Do you think it makes sense to add a `ReadOption` that allows early termination on a certain condition? This could help if you are representing the value as a set and just want to query value membership, you wouldn't need to retrieve the entire set if you've already found a merged entry that contains the value you're looking for.

Similarly, how would you implement a merge operator that removes a value from a list/set? I guess the user could do that in userspace by having special values that indicate "remove value X when merged" and just inserting that.

Reply:
> This is compatible with this RFC, so we can (and probably should) leave it as a follow up, but wanted to drop this note anyway. Do you think it makes sense to add a ReadOption that allows early termination on a certain condition? This could help if you are representing the value as a set and just want to query value membership, you wouldn't need to retrieve the entire set if you've already found a merged entry that contains the value you're looking for.

That sounds pretty interesting 👍 For example partial iterations (give me first / last 10 elements of the list) could become more feasible. I'll add it to list of possible followups.

> Similarly, how would you implement a merge operator that removes a value from a list/set? I guess the user could do that in userspace by having special values that indicate "remove value X when merged" and just inserting that.

Yes, it needs to happen in userspace. You're bit limited with associative operators in this regard though, non-associative ones are way more flexible for this case. Also `set` implementation is not a best example because you can create more performant version by each value being represented as a separate key with a shared prefix. (eg. key + hash)

Reply:
added